Cubical/Homotopy/WhiteheadsLemma.agda

```diff
@@ -0,0 +1,447 @@
+{-
+
+Whitehead's lemma for truncated spaces.
+
+-}
+{-# OPTIONS --safe #-}
+module Cubical.Homotopy.WhiteheadsLemma where
+
+open import Cubical.Foundations.Prelude
+
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Function
+open import Cubical.Foundations.GroupoidLaws
+open import Cubical.Foundations.HLevels
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Pointed
+open import Cubical.Foundations.Transport
+
+open import Cubical.Data.Nat
+open import Cubical.Data.Sigma
+open import Cubical.HITs.PropositionalTruncation hiding (map)
+open import Cubical.HITs.SetTruncation renaming (rec to sRec ; elim to sElim)
+open import Cubical.Homotopy.Group.Base
+open import Cubical.Homotopy.Loopspace
+
+private
+  variable
+    ℓ : Level
+
+-- lots of easy helper lemmas, some of which are probably duplicating lemmas
+-- proved elsewhere.
+private
+  SetTrunc→PropTrunc : {A : Type ℓ} → ∥ A ∥₂ → ∥ A ∥₁
+  SetTrunc→PropTrunc = sRec (isProp→isSet isPropPropTrunc) ∣_∣₁
+
+  isoPostComp : {A : Type ℓ} {a b c : A} (p : b ≡ c)
+    → Iso (a ≡ b) (a ≡ c)
+  Iso.fun (isoPostComp p) = λ r → r ∙ p
+  Iso.inv (isoPostComp p) = λ r → r ∙ (sym p)
+  Iso.rightInv (isoPostComp p) =
+    λ r → (sym (assoc r (sym p) p)) ∙ cong (r ∙_) (lCancel p) ∙ sym (rUnit r)
+  Iso.leftInv (isoPostComp p) =
+    λ r → (sym (assoc r p (sym p))) ∙ cong (r ∙_) (rCancel p) ∙ sym (rUnit r)
+
+  isoPostComp≃∙ : {A : Type ℓ} {a b c : A} {q : a ≡ b} (p : b ≡ c)
+    → ((a ≡ b) , q) ≃∙ ((a ≡ c) , q ∙ p)
+  fst (isoPostComp≃∙ p) = isoToEquiv (isoPostComp p)
+  snd (isoPostComp≃∙ p) = refl
+
+  isoPreComp : {A : Type ℓ} {a b c : A} (p : a ≡ b)
+    → Iso (b ≡ c) (a ≡ c)
+  Iso.fun (isoPreComp p) = λ r → p ∙ r
+  Iso.inv (isoPreComp p) = λ r → (sym p) ∙ r
+  Iso.rightInv (isoPreComp p) =
+    λ r → assoc p (sym p) r ∙ cong (_∙ r) (rCancel p) ∙ sym (lUnit r)
+  Iso.leftInv (isoPreComp p) =
+    λ r → assoc (sym p) p r ∙ cong (_∙ r) (lCancel p) ∙ sym (lUnit r)
+
+  isoPreComp≃∙ : {A : Type ℓ} {a b c : A} {q : b ≡ c} (p : a ≡ b)
+    → ((b ≡ c) , q) ≃∙ ((a ≡ c) , p ∙ q)
+  fst (isoPreComp≃∙ p) = isoToEquiv (isoPreComp p)
+  snd (isoPreComp≃∙ p) = refl
+
+  PathPCongLemma : {A B : Type ℓ}
+    (f : A → B)
+    (a a' : A)
+    (r : a ≡ a')
+    (b : B)
+    (p : f a ≡ b)
+    (q : f a' ≡ b)
+    → cong f r ≡ p ∙ (sym q)
+    → PathP (λ i → f (r i) ≡ b) p q
+  PathPCongLemma {B = B} f a a' =
```

Comment:
This is almost exactly `compPathR→PathP` from Foundations.Path. Maybe it can be skipped?